TODO/progressive-web-apps-with-react-js-part-3-offline-support-and-network-resilience.md

```diff
@@ -132,195 +103,95 @@ When a service worker controls a page, it can intercept each request being made
 
+记住你可以使用更简单的 SSR 或者 SPA 架构开发 PWA，但它没有同样的性能优势并且更依赖全页缓存。
 
+### 利用 Service Worker 启动低成本缓存
 
+这里列举两个用于不同离线场景的库：[sw-precache](https://github.com/GoogleChrome/sw-precache) 会自动事先缓存静态资源，[sw-toolbox](https://github.com/GoogleChrome/sw-toolbox) 处理运行时缓存以及回退策略。这两个库一起使用能达到互补的效果，需要提供静态内容外壳的性能策略时，总是从缓存中直接获取，而动态的或远程的资源则通过网络请求提供，需要时回退到缓存或静态响应里。
 
-We cache the shell offline using the Cache Storage API (via service worker) so that on repeat visits, the app shell can be loaded instantly so you get meaningful pixels on the screen really fast without the network, even if your content eventually comes from there.
-
-Note that you _can_ ship a PWA using a simpler SSR or SPA architecture, it just won’t have the same performance benefits and will instead rely more on full-page caching.
-
-### Low-friction caching with Service Worker Libraries
-
-Two libraries were developed to handle two different offline use cases: [sw-precache](https://github.com/GoogleChrome/sw-precache) to automate precaching of static resources, and [sw-toolbox](https://github.com/GoogleChrome/sw-toolbox) to handle runtime caching and fallback strategies. The libraries complement each other nicely, and allowed us to implement a performant strategy in which a static content “shell” is always served directly from the cache, and dynamic or remote resources are served from the network, with fallbacks to cached or static responses when needed.
-
-AppShell caching: Our static resources (HTML, JavaScript, CSS, and images) provide the core shell for the web application. Sw-precache allows us to make sure that most of these static resources are cached, and that they are kept up to date. Precaching every resource that a site needs to work offline isn’t feasible.
-
-Runtime caching: Some resources are too large or infrequently used to make it worthwhile, and other resources are dynamic, like the responses from a remote API or service. But just because a request isn’t precached doesn’t mean it has to result in a NetworkError. sw-toolbox gives us the flexibility to implement request handlers that handle runtime caching for some resources and custom fallbacks for others.
-
-_sw-toolbox supports a number of different caching strategies including__network-first_ _(ensure the freshest data is used if available, but fall back to the cache),_ _cache-first_ _(check a request matches a cache entry, fallback to the network),_ _fastest_ _(request resources from both the cache and network at the same time, respond with whichever returns first). It’s important to understand the_ [_pros and cons_](https://developers.google.com/web/fundamentals/instant-and-offline/offline-cookbook/) _of these approaches._
-
-
-
-
-
-
-
-
-
-
+应用外壳缓存：静态资源（HTML, JavaScript, CSS 和 images）提供 web 应用的核心外壳。Sw-precache 确保绝大多数这类静态资源都被缓存下来，并且保持更新。预缓存一个网站离线工作需要的所有资源显然是不现实的。
 
+运行时缓存：一些过于庞大或者很少使用资源，还有一些动态资源，像来自远程 API 或服务的响应。但是请求没有被事先缓存并不意味着它必须返回网络错误。sw-toolbox 让我们得以灵活实现请求的处理，这能够处理某些资源的运行时缓存和其他资源的自定义回退。
```

Comment:
But just because a request isn’t precached doesn’t mean it has to result in a NetworkError.

但是请求没有被事先缓存并不意味着它必须返回网络错误。

直接翻译有点怪，建议修改成： 没有被预先缓存的请求并不一定就是网络错误。